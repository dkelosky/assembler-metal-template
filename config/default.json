{
    "job": {
        "name": "TEMPLATE",
        "account": "#ACCT"
    },
    "program": {
        "name": "TEMPLATE"
    }
}