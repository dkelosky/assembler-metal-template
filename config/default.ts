import bindOptions from "./imports/bind/bindOptions";
import dataSets from "./imports/dataSets";
import uploads from "./imports/uploads";
import assemblyOptions from "./imports/assembly/assemblyOptions";
import assemblyMaclibs from "./imports/assembly/assemblyMaclibs";
import metalOptions64 from "./imports/metal/metalOptions64";
import metalIncludes from "./imports/metal/metalIncludes";
import metalOptions from "./imports/metal/metalOptions";

export default {

    // settings for all other sections
    settings: {
        hlq: "PUBLIC.TEMPLATE",
        name: "TEMPLATE",
        account: "#ACCT",
        description: "ASM/BIND/RUN",
        messageClass: "A",
        jobClass: "A",
    },

    // working data sets to allocate
    dataSets,

    // mapping local z/OS to LLQ data sets
    uploads,

    // job info
    job: {

        // metal c compilation configuration
        compile: {
            options: metalOptions64,
            includes: metalIncludes,
            sources: {
                TEMPLATE: {},
                AMS: {
                    options: metalOptions
                },
                AMSTEST: {},
            },
        },

        // assemble configuration
        assemble: {
            options: assemblyOptions,
            includes: assemblyMaclibs,
            sources: {
                TEMPLATE: {
                    // override options here
                },
                AMS: {},
                AMSTEST: {},
            },
        },

        // bind configuration
        bind: {
            options: bindOptions,
            // includes: nothing included in every bind right now
            sources: {
                TEMPLATE: {
                    includes: [
                        "AMS",
                    ]
                    // override options here
                },
                AMSTEST: {
                    includes: [
                        "AMS",
                    ]
                }
            }
        },

        // executables
        execute: {
            // options: no parms to every exec right now (options are PARM=)
            sources: {
                TEMPLATE: {
                    // override options here (options are PARM=)
                    options: "'HELLO WORLD'",
                },
                AMSTEST: {
                    options: "'HELLO WORLD'",
                },
            }
        }
    },

    convert: {
        assemble: {
            options: assemblyOptions,
            includes: assemblyMaclibs,
            sources: {
                DCBD: {
                    // override options here
                },
                IHADCBE: {},
                IHAECB: {},
                JFCB: {},
            },
        },
    },

    deploy: {
        ftp: {
            options: [
                "cd 'KELDA16.WORK.LOADLIB'",
                "mdelete *"
            ],
            includes: [
                "TEMPLATE",
            ],
            sources: {
                CA11: {},
            }
        }
    },
}