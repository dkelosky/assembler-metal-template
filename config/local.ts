export default {
    settings: {
        hlq: "KELDA16.METALC"
    },
    job: {
        name: "KELDA16$",
        account: "105300000",
    },
}