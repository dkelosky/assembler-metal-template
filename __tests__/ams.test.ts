process.env.ALLOW_CONFIG_MUTATIONS = "yes"; // value doesn't matter

import { execSync } from "child_process";
import * as config from "config";
import * as fs from "fs";
import * as handlebars from "handlebars";

const TIMEOUT = 20000;

const test = [{
        in: "hello world"
    },
    {
        in: "good bye world"
    }];

(config as any).test = test;

let buildResponse: any;
let testResponse: any;

describe("AMS tests", () => {

    beforeAll(() => {

        // build and run jcl to compile and bind all source
        const gen = execSync(`npm run genjcl:build`);
        console.log(gen.toString());
        let cmd = "zowe jobs submit lf \"./build/custom.jcl\" --directory \"./output\" --rfj"

        buildResponse = JSON.parse(execSync(cmd).toString());

        // build and run jcl to test service
        const jcl = fs.readFileSync("./zossrc/jcl/amstest.jcl").toString();
        const compiled = handlebars.compile(jcl);
        const rendered = compiled(config);

        if (!fs.existsSync("./build")) fs.mkdirSync("./build");
        fs.writeFileSync("./build/amstest.jcl", rendered);
        console.log("Generated custom JCL to ./build/amstest.jcl");

        cmd = "zowe jobs submit lf \"./build/amstest.jcl\" --directory \"./output\" --rfj"
        testResponse = JSON.parse(execSync(cmd).toString());

    }, TIMEOUT)

    describe("build tests", () => {
        it("should build without errors", () => {
            expect(buildResponse.data.retcode).toBe("CC 0000");
        });
    });

    describe("execution tests", () => {
        it("should read all input lines", () => {
            test.forEach((testData, index) => {
                const file = fs.readFileSync(`output/${testResponse.data.jobid}/RUN${index}/SYSPRINT.txt`).toString().trim();
                expect(file).toBe(testData.in);
            });
        });

        it("show our control block is built correctly", () => {
            const responses = fs.readFileSync(`output/${testResponse.data.jobid}/RUN0/SNAP.txt`).toString().split('\n');
            let index;
            responses.forEach((response, x) => {
                if (response.indexOf("Important Control Block") > -1) {
                    index = x + 1;
                }
            });

            expect(index).toBeDefined();

            if (index) expect(responses[index].substr(9, responses[index].length)).toMatchSnapshot();
        });
    });
});
