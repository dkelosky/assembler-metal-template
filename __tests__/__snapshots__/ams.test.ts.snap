// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AMS tests execution tests show our control block is built correctly 1`] = `"                            C3E3D9D3    0000000D 00000000 00000000            *            CTRL............    *"`;
